### PR TITLE
Add density compensation estimation code

### PR DESCRIPTION
**Pull request by chaithyagr**
Hello Andy!

I have written a brief code to allow estimation of density based on [Pipe et al](https://onlinelibrary.wiley.com/doi/abs/10.1002/%28SICI%291522-2594%28199901%2941%3A1%3C179%3A%3AAID-MRM25%3E3.0.CO%3B2-V) .

It seems to not be exactly going well with the overall flow, however, if we can figure it out, it will be a great addition to this library.

For now, what I have added is basically a method to carry out interpolation and gridding by early stopping.

Then, we can use the following method to estimate the density compensation as written in `pysap-mri` here: 

https://github.com/CEA-COSMIC/pysap-mri/blob/7b242672fef0398134b9e2f19af47cfc51ae8de1/mri/operators/fourier/utils.py#L313-L343

Let me know what you think of this codes for now. We can iterate through.

### CODE REVIEW

File: CUDA/CMakeLists.txt [hunk 1]
```diff
@@ -96,14 +96,12 @@ SET(FERMI_GPU OFF CACHE BOOL "Enable build for (old) Fermi architectures (Comput
 
 
 IF(FERMI_GPU)
-  set(MY_NVCC_FLAGS -gencode arch=compute_20,code=sm_20)
-  list(APPEND MY_NVCC_FLAGS -gencode arch=compute_30,code=sm_30)
+  set(MY_NVCC_FLAGS -gencode arch=compute_30,code=sm_30)
```
Comment (chaithyagr) [inline CUDA/CMakeLists.txt:99]: sm_20 is deprecated, i think this is why you have build issues with cuda 11 #89

File: CUDA/CMakeLists.txt [hunk 2]
```diff
@@ -96,14 +96,12 @@ SET(FERMI_GPU OFF CACHE BOOL "Enable build for (old) Fermi architectures (Comput
...
-  set(MY_NVCC_FLAGS -gencode arch=compute_20,code=sm_20)
-  list(APPEND MY_NVCC_FLAGS -gencode arch=compute_30,code=sm_30)
+  set(MY_NVCC_FLAGS -gencode arch=compute_30,code=sm_30)
   list(APPEND MY_NVCC_FLAGS -gencode arch=compute_50,code=sm_50)
   list(APPEND MY_NVCC_FLAGS -gencode=arch=compute_52,code=sm_52)
   list(APPEND MY_NVCC_FLAGS -gencode=arch=compute_52,code=compute_52)
 ELSE(FERMI_GPU)
-  set(MY_NVCC_FLAGS -arch=sm_30 -gencode arch=compute_30,code=sm_30)
-  list(APPEND MY_NVCC_FLAGS -gencode arch=compute_50,code=sm_50)
+  set(MY_NVCC_FLAGS -gencode arch=compute_50,code=sm_50)
```
Comment (chaithyagr) [inline CUDA/CMakeLists.txt:104]: Oh yes, even sm_30 was deprecated

File: CUDA/inc/gpuNUFFT_types.hpp [hunk 3]
```diff
@@ -188,7 +188,8 @@ enum GpuNUFFTOutput
   /** \brief Stop after iFFT step. */
   FFT,
   /** \brief Stop after deapodization step. */
-  DEAPODIZATION
+  DEAPODIZATION,
+  DENSITY_ESTIMATION,
```
Comment (chaithyagr) [inline CUDA/inc/gpuNUFFT_types.hpp:192]: A new tag for density estimation

**Review by chaithyagr — COMMENTED** [review]
A minute set of comments on codes

**Comment by andyschwarzl** [comment]
Hi @chaithyagr,

sorry for the late response.

Do you still have problems with the overall flow?

Best regards

**Comment by chaithyagr** [comment]
No no the flow works, it's just that it's not best code wise. You can give me some ideas too on this.

**Comment by andyschwarzl** [comment]
Sure my code is already very nested and hard to read, I'm sorry but do not have the time to make it clearer. I've mainly used the early-out options (GpuNUFFTOutput) for debugging purposes. It would be ok to integrate it as is, but maybe you keep an eye on freeing the device memory at least in your added code paths. Thanks!

**Comment by chaithyagr** [comment]
I have tested the codes about freeing, is there anything that I missed?

**Pull request merged**